process-engine/process_engine_core
# ✨ Implement join gateway discovery

## Changes

1. Add `findJoinGatewayAfterSplitGateway` to `ProcessModelFacade`, which discovers a matching Join Gateway for a given Split Gateway
    - Works for nested gateways of the same type, as well as for nested gateways of a different type
    - Works with branches that lead to a dead end (relevant for Exclusive Gateways, which can have branches ending in an EndEvent)
    - Ignores paths started by BoundaryEvents
    - Works for nested split gateways that have no corresponding join gateway (which can also happen with Exclusive Gateways)
2. When running ParallelSplitGateways, the matching Join Gateway is now discovered beforehand and an instance for it is registered at the ioc container
    - This ensures that no duplicated instances for the same Join Gateway are created
3. When starting multiple parallel running branches, a timeout of 33ms is waited between each branch
    - This is to help solve a problem with race conditions, when multiple branches arrive at the same ParallelJoinGateway at the exact same moment
    - See referenced issue for details about this problem
    - Note that this is only a workaround, we still need a viable solution to fix this issue permanently
4. Fix issue that caused internal ServiceTasks to crash, if only a single argument was provided

## Issues

Part of https://github.com/process-engine/process_engine_runtime/issues/463

PR: #320

## How to test the changes

- Run the diagram in the referenced issue several times in a row
- See that it will end as expected
